### PR TITLE
adds webrtc presence features

### PR DESCRIPTION
**Pull request by pretentious7**
is ready now

### CODE REVIEW

**Comment by pretentious7** [comment]
ready now.

File: js-pkg/apps/demos/src/pages/index.tsx [hunk 1]
```diff
@@ -32,6 +32,11 @@ const demos = [
     title: 'WebRTC Chat',
     href: '/webrtc',
     description: 'Simple chat with WebRTC datachannels, using DriftDB for signaling.'
+  },
+  {
+    title: 'WebRTC presence',
+    href: '/webrtc-canvas',
+    description: 'cursors with ghosts!'
```
Comment (rolyatmax) [inline js-pkg/apps/demos/src/pages/index.tsx:39]: Nice description haha but we should be clearer about exactly what we're trying to demonstrate with this demo and what that has to do with DriftDB.

Also, maybe `webrtc-presence` is a better href?

And capitalize both words in the title, like the other titles.

File: js-pkg/apps/demos/src/pages/webrtc-canvas.jsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+function drawCursor(ctx, userColor, name, mousePosition) {
+  ctx.beginPath()
+  ctx.fillStyle = userColor
+  ctx.moveTo(mousePosition[0], mousePosition[1])
+  ctx.arc(mousePosition[0], mousePosition[1], 5, 0, Math.PI * 2)
+  ctx.fill()
+  ctx.font = '12px monospace'
+  ctx.fillText(name, mousePosition[0] + 8, mousePosition[1] - 8)
+  ctx.fillText('SLDKFJS', mousePosition[0] + 8, mousePosition[1] - 8)
```
Comment (rolyatmax) [inline js-pkg/apps/demos/src/pages/webrtc-canvas.jsx:98]: What's this?

File: js-pkg/apps/demos/src/pages/webrtc-canvas.jsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    const boundingClientRects = event.currentTarget.getBoundingClientRect()
+    const x = event.clientX - boundingClientRects.x
+    const y = event.clientY - boundingClientRects.y
+    setMousePosition([x, y])
+  }
+
+  return (
+    <div>
+      <div className="flex items-center justify-between">
+        <h1 className="text-2xl font-bold text-gray-800">DriftDB - Shared Canvas Demo</h1>
```
Comment (rolyatmax) [inline js-pkg/apps/demos/src/pages/webrtc-canvas.jsx:56]: Need to retitle this

File: js-pkg/apps/demos/src/pages/webrtc-canvas.jsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  usePresence,
+  useWebRtcPresence,
+  DriftDBProvider
+} from 'driftdb-react'
+
+const HUE_OFFSET = (Math.random() * 360) | 0
+const USER_COLOR = randomColor()
+const RTC_COLOR = randomColor()
+
+function SharedCanvas() {
```
Comment (rolyatmax) [inline js-pkg/apps/demos/src/pages/webrtc-canvas.jsx:15]: Let's make this the name of the demo like we do in the other demos.

File: js-pkg/apps/demos/src/pages/webrtc-canvas.jsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+function drawCanvas(ctx, presence, webRtcPos) {
+  if (!ctx) return
+  ctx.clearRect(0, 0, ctx.canvas.width, ctx.canvas.height)
+
+  const userArr = [...webRtcPos.values(), ...Object.values(presence)]
+
+  userArr.map((user) => {
+    const { mousePosition, userColor, name } = user.value
+    drawCursor(ctx, userColor, name, mousePosition ?? [0, 0])
```
Comment (rolyatmax) [inline js-pkg/apps/demos/src/pages/webrtc-canvas.jsx:86]: Should probably only draw if `mousePosition !== null`.

File: js-pkg/apps/tests/package.json [hunk 6]
```diff
@@ -12,7 +12,7 @@
     "bun-types": "^0.5.7",
     "driftdb-internal": "*"
   },
-  "devDependencies": {
+  "optionalDependencies": {
```
Comment (rolyatmax) [inline js-pkg/apps/tests/package.json:15]: Why this change?

File: js-pkg/packages/driftdb-react/src/index.tsx [hunk 7]
```diff
@@ -315,6 +316,47 @@ export function useLatency(): number | null {
...
+  }
+  return {
+    send: (msg: string) => WebRtcBroadcastChannelRef.current!.send(msg),
+    setOnMessage: (onMessage: WebRtcOnMessage) =>
+      WebRtcBroadcastChannelRef.current!.setOnMessage(onMessage),
+    peers: WebRtcBroadcastChannelRef.current!.peers()
+  }
+}
+
+export function useWebRtcPresence(vals: any, throttle = 0) {
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb-react/src/index.tsx:337]: Since this is part of the public API, let's add an explicit return type along with a docstring comment.

File: js-pkg/packages/driftdb-react/src/index.tsx [hunk 8]
```diff
@@ -315,6 +316,47 @@ export function useLatency(): number | null {
...
+    }
+  }, [peers])
+  React.useEffect(() => {
+    let count = 0
+    setOnMessage((msg) => {
+      rtcMap.current!.set(msg.sender, msg)
+      setMessageCount(count++)
+    })
+  }, [messageCount])
+  return rtcMap.current
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb-react/src/index.tsx:357]: I think it makes sense to mirror the interface of `usePresence` by having this return an Object.

File: js-pkg/packages/tsconfig/react-library.json [hunk 9]
```diff
@@ -4,7 +4,7 @@
   "extends": "./base.json",
   "compilerOptions": {
     "jsx": "react-jsx",
-    "lib": ["ES2015", "dom"],
+    "lib": ["ES2017", "dom"],
```
Comment (rolyatmax) [inline js-pkg/packages/tsconfig/react-library.json:7]: Why was this change necessary?

File: js-pkg/packages/driftdb-react/src/index.tsx [hunk 10]
```diff
@@ -315,6 +316,47 @@ export function useLatency(): number | null {
...
+  const { send, setOnMessage, peers } = useWebRtcBroadcastChannel(throttle)
+  const rtcMap = useRef<Map<string, DataChannelMsg>>()
+  if (!rtcMap.current) rtcMap.current = new Map()
+  const [messageCount, setMessageCount] = useState(0)
+  React.useEffect(() => {
+    send(JSON.stringify(vals))
+  }, [vals])
+  React.useEffect(() => {
+    for (const peer of rtcMap.current!.keys()) {
+      if (![...peers].includes(peer)) rtcMap.current!.delete(peer)
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb-react/src/index.tsx:347]: Is the `[...peers]` necessary? Why not just `peers.includes(peer)`?

File: js-pkg/packages/driftdb-react/src/index.tsx [hunk 11]
```diff
@@ -315,6 +316,47 @@ export function useLatency(): number | null {
...
+    }
+  }, [peers])
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb-react/src/index.tsx:349]: Does building the presence object (or map, in this case) need to happen in a `useEffect`? I think this can just happen in the main function.

File: js-pkg/packages/driftdb-react/src/index.tsx [hunk 12]
```diff
@@ -315,6 +316,47 @@ export function useLatency(): number | null {
...
+  React.useEffect(() => {
+    let count = 0
+    setOnMessage((msg) => {
+      rtcMap.current!.set(msg.sender, msg)
+      setMessageCount(count++)
+    })
+  }, [messageCount])
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb-react/src/index.tsx:356]: I don't think I'm following the logic for why we need to set a new `onMessage` for the WebRTCBroadcastChannel for every message that comes in 🤔

File: js-pkg/packages/driftdb-react/src/index.tsx (hunk 10 above)
Comment (pretentious7) [reply js-pkg/packages/driftdb-react/src/index.tsx:347]: peers returns an iterator, so doesn't have an includes.

File: js-pkg/apps/tests/package.json (hunk 6 above)
Comment (pretentious7) [reply js-pkg/apps/tests/package.json:15]: I'm not sure. for whatever reason, the bun installscript doesn't seem to run correctly and causes npm to fail.

File: js-pkg/packages/driftdb-react/src/index.tsx (hunk 11 above)
Comment (pretentious7) [reply js-pkg/packages/driftdb-react/src/index.tsx]: I thought the peers thing would return the same reference, this was not correct.

File: js-pkg/packages/tsconfig/react-library.json [hunk 13]
```diff
@@ -4,7 +4,7 @@
   "extends": "./base.json",
   "compilerOptions": {
     "jsx": "react-jsx",
-    "lib": ["ES2015", "dom"],
+    "lib": ["ES2016", "dom"],
```
Comment (pretentious7) [inline js-pkg/packages/tsconfig/react-library.json:7]: note this is needed for array.includes

**Comment by pretentious7** [comment]
also switched over to 18.5.0, gets the package-lock diff way down.

File: js-pkg/packages/driftdb-react/src/index.tsx [hunk 14]
```diff
@@ -315,6 +316,56 @@ export function useLatency(): number | null {
...
+  }
+  const [_, setMessageCount] = useState(0)
+  React.useEffect(() => {
+    send(JSON.stringify(value))
+  }, [value])
+  React.useEffect(() => {
+    let count = 0
+    setOnMessage((msg) => {
+      rtcMap.current[msg.sender] = { value: msg.value, lastSeen: msg.lastSeen }
+      setMessageCount(count++)
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb-react/src/index.tsx:362]: It seems like what we really want is for changes to `rtcMap` to trigger rerenders. In that case, I'd suggest we use `useState` for the rtcMap instead of a ref. It looks like that's how `usePresence` works.

```ts
const [rtcPresence, setRtcPresence] = useState<Record<string, WrappedPresenceMessage<T>>>({})
// then we drop the messageCount stuff
React.useEffect(() => {
    setOnMessage((msg) => {
      rtcPresence[msg.sender] = { value: msg.value, lastSeen: msg.lastSeen }
      setRtcPresence(rtcPresence)
    })
})
```

Something like that?

Also, why can't we just pass this callback directly to the `useWebRtcBroadcastChannel()` instead of having it return a setter that we use right away? Did we already talk through this but I've just forgotten?

**Review by paulgb — COMMENTED** [review]
One comment, otherwise I think this is looking ready!

File: js-pkg/packages/driftdb/src/webrtc.ts [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+    cb1(elem)
+  }
+  for (const elem of difference(b, a)) {
+    cb2(elem)
+  }
+}
+
+export class SyncedWebRTCConnections extends WebRTCConnections {
+  presence: PresenceListener<string>
+  #peersToLastMsg: Record<string, WrappedPresenceMessage<any>> = {}
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb/src/webrtc.ts:99]: Can we use `private` like we do elsewhere?

File: js-pkg/packages/driftdb/src/webrtc.ts [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+    this.presence.updateState(newId)
+  }
+
+  sync(newPeers: Syncable<string>) {
+    newPeers = new Set(newPeers) //since this is used twice it can't be an iterable
+    sync(
+      {
+        [Symbol.iterator]: () => Object.keys(this.#peersToLastMsg).values(),
+        has: Reflect.has.bind({}, this.#peersToLastMsg)
+      },
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb/src/webrtc.ts:136]: I gotta say, I don't think this whole `Syncable` business is paying off. It makes the code a lot more complicated in several places, has weird edge cases you end up having to deal with (e.g. above, having to turn newPeers into a Set anyway because it can't be used twice as an iterable), and generally just adds a lot of stuff that gets in the way of understanding what's really happening. There wasn't really anything wrong with the way it was before, as I recall. Can we just go back to that?

Comment (pretentious7) [reply js-pkg/packages/driftdb/src/webrtc.ts:136]: I changed it cause that allocated a bunch every time sync was called, which was every single usePresence onMessage

Comment (pretentious7) [reply js-pkg/packages/driftdb/src/webrtc.ts:136]: but yea I guess I'm still allocating a set here, sigh, I'll change it back.

File: js-pkg/packages/driftdb/src/webrtc.ts [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+  sync(newPeers: string[]) {
+    this.setConnMap(
+      newPeers.map((peer) => [peer, this.getConn(peer) ?? this.addNewConnection(peer)])
+    )
+    this.peersToLastMsg = Object.fromEntries(
+      Object.entries(this.peersToLastMsg).reduce(
+        (a, b) => (newPeers.includes(b[0]) ? [...a, b] : [...a]),
+        [] as [string, WrappedPresenceMessage<any>][]
+      )
+    )
```
Comment (rolyatmax) [inline js-pkg/packages/driftdb/src/webrtc.ts:138]: Minor comment: there might be a simpler (and more performant!) way to write this.

Three things about this:

1. Creates a lot of unnecessary objects that are just thrown away by the end of the fn (an array when converting newPeersToLastMsg, an array for the reduce init value, and an array for every key in peersToLastMsg)
2. Uses includes() which isn't es2015
3. Is just hard to read 😬 

Consider this option, which is far more boring, and also doesn't suffer from the above:

```ts
const newPeersToLastMsg: Record<string, WrappedPresenceMessage<any>> = {}
for (const peer of newPeers) {
  newPeersToLastMsg[peer] = this.peersToLastMsg[peer]
}
this.peersToLastMsg = newPeersToLastMsg
```

Or this, which reuses the same object over and over:

```ts
for (const peer in this.peersToLastMsg) {
  if (!this.connMap.has(peer)) {
    delete this.peersToLastMsg[peer]
  }
}
```

**Review by rolyatmax — APPROVED** [review]
